File: swift/Sources/ImogenSDK/ImogenClient.swift
import Foundation

#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif

/// The imogen client.
///
/// ```swift
/// let imogen = ImogenClient(baseURL: "https://photos.example.com", token: "…")
/// let page = try await imogen.assets.list(AssetQuery(q: "harbour", limit: 50))
/// ```
///
/// In a context served by imogen itself, omit the token: the session cookie is enough.
public struct ImogenClient: Sendable {
    public let http: HTTPClient
    public let assets: Assets
    public let albums: Albums
    public let admin: Admin
    public let auth: Auth
    public let vault: Vault
    public let people: People

    public init(options: ClientOptions) {
        let http = HTTPClient(options: options)
        self.http = http
        self.assets = Assets(http: http)
        self.albums = Albums(http: http)
        self.admin = Admin(http: http)
        self.auth = Auth(http: http)
        self.vault = Vault(http: http)
        self.people = People(http: http)
    }

    public init(baseURL: String, token: String? = nil, session: URLSession = .shared) {
        self.init(
            options: ClientOptions(
                baseURL: baseURL,
                token: token.map { value in { @Sendable in value } },
                session: session
            )
        )
    }

    public var baseURL: String { http.baseURL }

    /// Confirms the server is reachable and reports its version.
    public func health() async throws -> Health {
        try await http.request("GET", "/api/v1/health")
    }
}

File: kotlin/src/main/kotlin/com/imogen/sdk/ImogenClient.kt
@file:Suppress("unused")

package com.imogen.sdk

/**
 * The imogen client.
 *
 * ```kotlin
 * val imogen = ImogenClient("https://photos.example.com", token = "…")
 * val page = imogen.assets.list(AssetQuery(q = "harbour", limit = 50))
 * ```
 *
 * In a context served by imogen itself, omit the token: the session cookie is enough.
 */
class ImogenClient(options: ClientOptions) : AutoCloseable {
    val http: HttpClient = HttpClient(options)
    val assets: Assets = Assets(http)
    val albums: Albums = Albums(http)
    val admin: Admin = Admin(http)
    val auth: Auth = Auth(http)
    val vault: Vault = Vault(http)
    val people: People = People(http)
    val pairing: Pairing = Pairing(http)

    constructor(baseUrl: String, token: String? = null) : this(
        ClientOptions(baseUrl = baseUrl, token = token?.let { { it } })
    )

    val baseUrl: String get() = http.baseUrl

    /** Confirms the server is reachable and reports its version. */
    suspend fun health(): Health = http.request("GET", "/api/v1/health")

    override fun close() = http.close()
}
